Support non-Tensor args in the Pipe API

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57325 Add `NoChunk` wrapper for pipeline args.
* **#57226 Support non-Tensor args in the Pipe API**

As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

pritamdamania committed:
Support non-Tensor args in the Pipe API

As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-Tensor args in the Pipe API"


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-Tensor args in the Pipe API"


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-Tensor args in the Pipe API"


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-Tensor args in the Pipe API"


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-Tensor args in the Pipe API"


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]

SciPioneer commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -21,77 +21,82 @@
+    An abstraction representing a microbatch in the pipeline.
+    """
+
+    def __init__(self, values) -> None:
```
Nit: can you add the type hint for `values`?

SciPioneer commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -21,77 +21,82 @@
+        """
+        if self.atomic:
+            return self._values.device
+        for value in self._values:
```
Will all the tensors here always share the same device?

SciPioneer commented on torch/distributed/pipeline/sync/microbatch.py:
```diff
@@ -21,77 +21,82 @@
+        """Retreives the underlying values for the batch"""
+        return self._values
+
+    def find_tensor_idx(self):
```
What's the purpose of this function? Will all the tensors be placed contiguously in `values_`?

rohan-varma replied:
it looks like we call this a few different times, is it worth caching? maybe a `property` decorator and a getter that only sets the value once if it has not already been set?

SciPioneer commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -375,9 +375,8 @@ def forward(self, *inputs) -> RRef:
-        there's type restriction. Input and output have to be a
-        :class:`~torch.Tensor` or a sequence of tensors. This restriction is
-        applied at partition boundaries too.
+        there's type restriction. Input and output have to contain atleast one
```
typo: at least

SciPioneer commented on torch/distributed/pipeline/sync/checkpoint.py:
```diff
@@ -99,31 +100,31 @@ def __init__(self, function: Function, batch: Batch) -> None:
-        input = tuple(self.batch)
+        inputs = tuple(self.batch)
 
-        # batch[0] is always requiring grad, because it has been passed
```
Shall we keep the comments here? Only need to change `batch[0]` to `batch[tensor_idx]` I guess.

SciPioneer commented on test/distributed/pipeline/sync/test_pipe.py:
```diff
@@ -381,22 +381,124 @@ class NonTensorTuple(nn.Module):
+    assert torch.allclose(d, res[6])
+
+    # Test one of the tensors can be None
+    res = model(a, b, c, None).local_value()
```
Can you also make `b` as a tensor? This way you also test 1) multiple tensor args; 2) a tensor arg does not have to be the last arg.

wayi1 reviewed (approved):
Thanks for extending the pipe API

rohan-varma commented on test/distributed/pipeline/sync/test_pipe.py:
```diff
@@ -381,22 +381,124 @@ class NonTensorTuple(nn.Module):
+
+    # Test one of the tensors can be None
+    res = model(a, b, c, None).local_value()
+    assert 5, len(res)
```
won't this just check if 5 is always true? should it be assert 5 == len(res)? or does pytest do something special with the assert keyword?

rohan-varma commented on test/distributed/pipeline/sync/test_pipe.py:
```diff
@@ -381,22 +381,124 @@ class NonTensorTuple(nn.Module):
+    model = Pipe(model)
+
+    with pytest.raises(TypeError):
+        model("hello", True)
```
add a comment that the failure is because we  Need atleast one tensor? also can we enhance the error msg to indicate this?

rohan-varma commented on torch/distributed/pipeline/sync/checkpoint.py:
```diff
@@ -302,32 +312,49 @@ def forward(
+        else:
+            tensors = []
+            for input in enumerate(inputs):
+                if torch.is_tensor(input):
```
doesn't enumerate always return a tuple? So this is always false, or am I missing something?

rohan-varma commented on torch/distributed/pipeline/sync/checkpoint.py:
```diff
@@ -302,32 +312,49 @@ def forward(
+        device = None
+        for input in inputs:
+            if torch.is_tensor(input):
+                device = input.device
```
currently device will be the device of the last input tensor - can we break when we populate device the first time? Is there any issue if they're on different devices?

pritamdamania committed:
Update on "Support non-Tensor args in the Pipe API"


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support non-Tensor args in the Pipe API"


As per the design outlined in
https://github.com/pytorch/pytorch/issues/53952, this PR adds support for
non-Tensor args in the pipeline.

The `NoChunk` wrapper hasn't been implemented yet and will be implemented in a
follow up PR.

Differential Revision: [D28083564](https://our.internmc.facebook.com/intern/diff/D28083564/)

[ghstack-poisoned]
